[bc-breaking] reference option for linear produce a pattern instead of reference linear module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62215
* #61942
* __->__ #61892
* #61859

Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

jerryzh168 committed:
[bc-breaking] reference option for linear produce a pattern instead of reference linear module

Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/nn/qat/modules/linear.py:
```diff
@@ -49,3 +50,11 @@ def from_float(cls, mod):
         qat_linear.bias = mod.bias
         return qat_linear
+
+    def to_float(self):
```
given @HDCharles recent work, might be good to copy over the right device and `is_traning` setting to the new module

jerryzh168 replied:
I think we already take care of the device when we copy the weight? since it's used when we initialize thew weight: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/linear.py#L81

jerryzh168 replied:
will add is_training

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

vkuzo commented on torch/nn/intrinsic/qat/modules/linear_relu.py:
```diff
@@ -36,3 +37,12 @@ def forward(self, input):
         return super(LinearReLU, cls).from_float(mod)
+
+    def to_float(self):
+        linear = torch.nn.Linear(self.in_features, self.out_features)
```
should we respect the device and is_training settings when creating the new module?

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[bc-breaking] reference option for linear produce a pattern instead of reference linear module"


Summary:
This PR changes is_reference=True for linear to produce a pattern consists of dequant - float linear - quant instead of reference linear module, this is useful for future transformations to custom backends, it is also helpful to simplify the implementation for
convert in the future.

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29810657](https://our.internmc.facebook.com/intern/diff/D29810657)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29810657).

ejguan commented:
Reverting this PR due to it breaks mypy. I think the error has already been shown in the test result of this PR.
Please do not bypass it when landing as mypy should not be a flaky test. It would break the experience for other PyTorch developers. Thank you.
